ls コマンドを作るオブジェクト指向版を作成しました

yumiya-g opened the pull request.

yumiya-g committed:
Add: 動作テスト用のダミーファイルを追加しました

yumiya-g committed:
Add: テストファイルを追加しました

- CommandクラスとFileクラス用

yumiya-g committed:
Fix: テスト用ダミーファイルの変更と、テストファイルの修正を行いました

yumiya-g committed:
Add: `-l`オプションなしで、ファイル名を3列表示で出力する

yumiya-g committed:
Fix: `File`クラスを修正しました

- ハードリンクとブロック数をハッシュに追加
- 更新日時を文字列にしてpaddingを調整

yumiya-g committed:
Fix: `l`オプション追加時の出力処理を実装しました

yumiya-g committed:
Update: テストファイルを修正しました

yumiya-g committed:
Fix: Lsクラスを修正しました

- 不要なコードを削除

yumiya-g committed:
Fix: Commandクラスを修正しました

- `l`オプションを持っている場合の条件分岐処理を削除
- `a`オプションを持っている場合の条件分岐処理を追加
- 不要なコードやコメントを削除

yumiya-g committed:
Fix: Fileクラスを修正しました

- インスタンス変数名は並び順を調整
- ファイル情報の整形処理を修正
- オプションによる条件分岐処理を削除
- 不要なコードやコメント等を削除

yumiya-g committed:
Fix: Rowクラスを修正しました

- `l`と`r`オプションを持っている場合の条件処理を追加
- `l`の有無で、ファイル名またはファイル情報を出力する処理を切り替える
- 不要なコードやコメント等を削除

yumiya-g committed:
Delete: テスト用のダミーファイルを削除しました

yuuu commented on 08.ls_object/command.rb:
```diff
@@ -0,0 +1,41 @@
+    opt.on('-l') { |v| v }
+    options = opt.getopts(argv).keys.map(&:to_sym)
+    directory_name = opt.parse(argv)
+    { directory_name: directory_name, options: options }
```
Ruby 3.1以降であれば `{ directory_name:, options: }` と記述できますね。

yumiya-g replied:
キー名と値が同じ場合は省略記法が使えるので、修正しました🙏
https://blog.netlab.jp/2021/12/20/hash-value-omission.html

yuuu commented on 08.ls_object/row.rb:
```diff
@@ -0,0 +1,59 @@
+
+MAX_COLUMNS = 3
+
+class Row
```
`Row` というクラス名なので、このクラスのインスタンスはlsコマンドの出力のうちの1行分を担うように見えますが、実際には全行を担っていますね。全行を担うことがわかる名前に変えるか、 `Row` という名前に合う処理に変えるといった対応をお願いします 🙏

yumiya-g replied:
ファイル情報をリストで出力しているので`FileList`に名前と処理を変更しました。

yuuu commented on 08.ls_object/command.rb:
```diff
@@ -0,0 +1,41 @@
+    files = options.include?(:a) ? Dir.glob(dir_name, ::File::FNM_DOTMATCH) : Dir.glob(dir_name)
+
+    files.map do |file|
+      Ls::File.new(file, options).generate_file_stats
```
この時点で `.generate_file_stats` しないほうが、後々の仕様追加に対応しやすいのではないでしょうか？

以下ドキュメントに書いてある話と同じだと思っています。
https://bootcamp.fjord.jp/pages/utilize-objects

yumiya-g replied:
`generate_file_stats`メソッドを使わず、オブジェクトのまま出力して他のクラスに渡せるように調整を行いました。

yuuu commented on 08.ls_object/command.rb:
```diff
@@ -0,0 +1,41 @@
+
+  def fetch_files
+    dir_name = directory_name.empty? ? '*' : "#{directory_name.first}/*"
+    files = options.include?(:a) ? Dir.glob(dir_name, ::File::FNM_DOTMATCH) : Dir.glob(dir_name)
```
ここは第2引数の有無が異なるだけなので、コードの重複を減らせないか検討をお願いします 🙏

yumiya-g replied:
`if`分で`a`オプションを持つ時に第2引数を追加するよう変更しました🙏

yuuu commented on 08.ls_object/file.rb:
```diff
@@ -0,0 +1,73 @@
+
+module Ls
+  class File
+    attr_reader :file_stats, :date, :permission
```
ここはおそらくテストのためだけに `attr_reader` しているように見えます。 `generate_file_stats` を使って値を取り出すようにすれば、ここは完全にカプセル化できるのではないでしょうか？

yumiya-g replied:
`attr_reader`から、テストのために使用していた`:date`と` :permission`を削除しました。
`file_stats`は別クラスから呼び出すため、残しておきました。

yuuu commented on 08.ls_object/file.rb:
```diff
@@ -0,0 +1,73 @@
+    attr_reader :file_stats, :date, :permission
+
+    def initialize(file = nil, options = [])
+      @options = options
```
ここも全体的にプリミティブな値をインスタンス変数として持つようになっていますが、 `file` と `options` を持つようにして、パブリックメソッドが呼ばれたタイミングで所望の形に変更するようにしたほうが、読みやすく拡張性が高くなるはずです。

yumiya-g replied:
ここではインスタンス変数`@file_name`を設定して、`generate_file_stats`メソッドでファイル情報を変換するように調整しました。

yuuu commented on 08.ls_object/row.rb:
```diff
@@ -0,0 +1,59 @@
+  def initialize(file_stats, options = [])
+    @options = options
+    @file_stats = @options.include?(:r) ? file_stats.reverse.compact : file_stats.compact
+    options.include?(:l) ? display_stats : display_name
```
このクラスのメソッドは「 `display_stats` のためのメソッド」と「 `display_name` のためのメソッド」がくっきり分かれていますね。これはオブジェクト指向用語で「凝集度が低い」状態です。

`display_stats` と `display_name` という関心が異なるものが1つのクラスに共存しているので、これらを別々のクラスにできないか検討をお願いします 🙏

yumiya-g replied:
出力を`FileName`クラスと`FileStat`クラスに分割しました。これらのクラスは出力データを整形するため、関数的に使用しているのでクラスメソッドで定義しています。
[【プログラミング初心者向け】クラスメソッドとインスタンスメソッドはどう使い分けるべき？](https://blog.jnito.com/entry/2022/07/20/080512)

yuuu commented on 08.ls_object/test/file_test.rb:
```diff
@@ -0,0 +1,33 @@
+                 }
+  end
+
+  def test_generate_date_object
```
メソッドが分かれているのでテストも書いた、ということだと思いますが、ここは `test_generate_stats` で一緒にテストできているので、必ずしもテストは必要ないと思いました。もちろんテストはあればあるほどよいので、このままでもOKです。

yumiya-g replied:
こちらはプライベートメソッドをテストしているため、不要かとも思ったのですがそのままにしておきます🙏

yuuu commented on 08.ls_object/test/command_test.rb:
```diff
@@ -0,0 +1,40 @@
+require 'minitest/autorun'
+require_relative '../command'
+
+class CommandTest < Minitest::Test
```
Commandの責務は渡したディレクトリ名・オプションを持つことの他に `file_stats` を適切に集めるというものがありますが、現状のテストではそこを確認できていないですね。

yumiya-g replied:
`test_file_stats_exist`メソッドにて、`file_stats`が取得できているか確認するよう調整しました。

yumiya-g committed:
Update: ハッシュを省略記法に変更しました

yumiya-g committed:
Update: 隠しファイル取得の条件分岐処理を見直しました

yumiya-g committed:
Fix: ファイル名を取得とLs::Fileクラスオブジェクトの処理を修正しました

- インスタンス変数名を変更
- `-a`オプションの有無で隠しファイルを取得する
- Ls::Fileクラスのオブジェクトをそのまま保持する
  - 第2引数が不要だったので削除

yumiya-g committed:
Rename: `Row`クラスを`FileList`クラスに変更しました

- `Row`クラスが持っていた出力を`FileStat`クラスと`FileName`クラスに分割
- `FileStat`クラスと`FileName`クラスはクラスメソッドで定義

yumiya-g committed:
Fix: インスタンス変数の持ち方を変更して、パプリックメソッドでプリミティブな値に変換しました

yumiya-g committed:
Add: ファイル名の出力と、ファイル情報の出力を別クラスにしました

yumiya-g committed:
Fix: テストファイルを修正しました

yumiya-g committed:
Delete: `Debug`Gemを削除しました

yuuu commented on 08.ls_object/command.rb:
```diff
@@ -0,0 +1,42 @@
+
+  def fetch_files
+    directory_path = [directory_name.empty? ? '*' : "#{directory_name.first}/*"]
+    directory_path << ::File::FNM_DOTMATCH if options.include?(:a)
```
ここで挿入している情報は `directory_path` ではないので名前の見直しをお願いします 🙏

yumiya-g replied:
こちらは引数が入るので`directory_path`は適切ではなかったですね。`Dir.glob`の引数が入るので`arguments`に変更しました。

yumiya-g committed:
Fix: `Dir.glob`に渡す変数名を見直しました

yumiya-g committed:
Fix: ファイル名とファイルの詳細情報を出力するクラス名を見直しました

yumiya-g committed:
Fix: `self`メソッドを削除してインスタンスメソッドでデータを出力する処理に変更しました

yumiya-g commented on 08.ls_object/file_list.rb:
```diff
@@ -0,0 +1,19 @@
+  private
+
+  def display_files(options)
+    options.include?(:l) ? FileStat.display(@files) : FileName.display(@files, MAX_COLUMNS)
```
マニュアルを確認したところ"ディレクトリの内容"というキーワードから、クラス名を`DirectoryContent`と`DirectoryContentLongFormat `に変更しました。

yuuu replied:
これらのクラスは `-l` またはそうでない時に使われるものなので、 `-l` の部分の文章から単語を選定したいです。
ちなみに `-l` を指定していないときの挙動は `-C` を指定したときと同じなので、 `-C` の文章も併せて確認してみてください。

yumiya-g replied:
マニュアルより、オプション指定がない場合は`-C`であることを確認いたしました🙏
再度クラス名を検討して修正を行いました。

yumiya-g commented on 08.ls_object/file_name.rb:
```diff
@@ -0,0 +1,22 @@
+# frozen_string_literal: true
+
+class FileName
+  class << self
```
`self`メソッドを使うとクラスに依存したインスタンスが生成されることがわかりましたので、インスタンスメソッドで呼び出すように修正しました。

yuuu commented on 08.ls_object/directory_content.rb:
```diff
@@ -0,0 +1,20 @@
+# frozen_string_literal: true
+
+class DirectoryContent
+  def display(files, columns)
```
`files` は `display` メソッドの引数とするのではなく、 `initialize` 
の引数としてインスタンス変数に覚えるようにしてはどうでしょう？

オブジェクト指向は、「クラスの中にデータとメソッドを閉じ込めることでセットで扱えるようにする」というのが根底にあります。

このメソッドの引数の `files` はこのメソッドだけで使うものではなく、このクラス内のあらゆるメソッドで使うものではないでしょうか？(今はメソッドが1つしかないのでわかりにくいかもしれませんが)

yumiya-g replied:
仰るとおり、メソッドが1つしか無いので`initialize`メソッドでインスタンス変数を作成していませんでした。
仮に今後メソッドが増えることを想定するとクラス内のメソッドで使うインスタンス変数を用意したほうがよいですね🙏

yumiya-g committed:
Fix: ファイル情報の出力処理を修正しました

- `initialize`メソッドでインスタンス変数を作成
- `display`メソッドを`private`に設定

yumiya-g committed:
Fix: ファイル情報を出力するクラス名を変更しました

yuuu commented on 08.ls_object/long_format_output.rb:
```diff
@@ -0,0 +1,39 @@
+class LongFormatOutput
+  def initialize(files)
+    @files = files
+    display
```
今は `display` しかpublicなメソッドがないのでここで呼んでしまっても違和感がないかもしれませんが、別のメソッドを追加するシチュエーションが生じたときに困ってしまわないでしょうか？

例えば `initialize` 後に別途ファイルを追加するメソッドを追加する、などは起こり得ると思います。

yumiya-g replied:
`initialize`メソッドではなく、インスタンスメソッドで`display`を呼び出すように酒修正を行いました🙏

yuuu commented on 08.ls_object/file_list.rb:
```diff
@@ -0,0 +1,19 @@
+
+MAX_COLUMNS = 3
+
+class FileList
```
ここまで小さくなるとこのクラスの責務はほとんどなく、存在することでむしろ全体を複雑化しているように見えます 👀 

`Command` クラスと統合してはどうでしょう？

yumiya-g replied:
オプションで分岐するだけになっている`FileList`クラスの処理を`Command`クラスに統合しました。

yuuu commented on 08.ls_object/default_output.rb:
```diff
@@ -0,0 +1,26 @@
+# frozen_string_literal: true
+
+class DefaultOutput
```
この名前だと、「lsコマンドの出力形式のデフォルトを `-C` から `-l` に変更する」という仕様変更が発生した場合に、このクラス名も変更する必要が出てしまいます。

「デフォルト」というのはこのクラスの本質ではないので、名前を見直してください。繰り返しになりますが `-C` のmanページを読んでみることをおすすめします。

yumiya-g replied:
manコマンドの説明より`MultiColumnOutput`と命名しました。

```
-C      Force multi-column output; this is the default when output is to
             a terminal.
```

yumiya-g committed:
Fix: `l`オプションなしでファイル名を出力するクラス名を変更しました

yumiya-g committed:
Fix: ファイル情報をインスタンスメソッド`display`で出力するよう調整しました

yumiya-g committed:
Delete: `FileList`クラスを削除して、`Command`クラスに統合しました

yuuu commented on 08.ls_object/command.rb:
```diff
@@ -0,0 +1,46 @@
+
+  def display_files
+    files = options.include?(:r) ? @files.reverse : @files
+    options.include?(:l) ? LongFormatOutput.new(files).display : MultiColumnOutput.new(files, MAX_COLUMNS).display
```
`.display` が重複しているので、三項演算子の外側に出すことで重複を解消しましょう。
オブジェクト指向におけるポリモーフィズム、ダックタイピングと呼ばれる考え方が、まさにここでやりたいことに該当します 👍

yumiya-g committed:
Fix: 三項演算子で重複していた`display`メソッドを解消しました
